Updating documentation (#58)
canonical/grafana-agent-k8s-operator: INTEGRATING.md

# Integrating Grafana agent

Grafana Agent Charmed Operator can be integrated with other charms through the following relations.

## Provides

### Logging provider

```yaml
  logging-provider:
    interface: loki_push_api
```

Grafana Agent Charmed Operator may receive logs from any charm that supports the [`loki_push_api`](https://charmhub.io/loki-k8s/libraries/loki_push_api) relation interface.

> Besides, the `logging-provider` relation can inject and setup `Promtail` into the workload container of the consumer side of the relation.

Author: Good catch!. This paragraph is wrong. Thanks!

The information exchanged through the `loki_push_api` interface can be broken down into two parts:

  - Grafana Agent charm provides an [endpoint URL](https://grafana.com/docs/loki/latest/api/#post-lokiapiv1push) to receive log from [Loki clients](https://grafana.com/docs/loki/latest/clients/) that relates with this charm.
  - Grafana Agent charm may also receive alert rules which tell when to raise alerts. These rules are read from a directory named `loki_alert_rules`, if present at the top level, within the client charm's source (`src`) directory.

For instance let's say that we have a [Zinc Charmed Operator](https://charmhub.io/zinc-k8s) that implements the other side (`requires`) of the relation.

> After deploying this charm, we can relate `Grafana Agent` and `Zinc` through `loki_push_api` relation interface:

Reviewer: Need to clearly differentiate between the charm's function as provider and as requirer of the same relation, but how about:

> To have a charm send logs to loki, you can use one of the "consumer" object offered by the `loki_push_api`.
> 
> First, specify the relation in the charm's metadata.yaml:
> 
> ```yaml
> requires:
>   logging:
>     interface: loki_push_api
> ```
> 
> Next, obtain the library from charmhub:
> 
> ```shell
> charmcraft fetch-lib charms.loki_k8s.v0.loki_push_api
> ```
> 
> Then, use the charm library in the charm:
> 
> ```python
> self._logging = LogProxyConsumer(...)
> # or
> self._logging = LokiPushApiConsumer(...)
> ```
> 
> Finally, relate the app (e.g. zinc) to loki:

Author: I like it!

```bash
juju relate grafana-agent-k8s:logging-provider zinc-k8s
```

Once the relation is established, Zinc charm can start sending logs to Grafana Agent charm.

### Grafana dashboard

```yaml
  grafana-dashboard:
    interface: grafana_dashboard
```

Over the `grafana-dashboard` relation using the [`grafana-dashboard`](https://charmhub.io/grafana-k8s/libraries/grafana_dashboard) interface, this Grafana Agent charm also provides meaningful dashboards about its metrics to be shown in a [Grafana Charm ](https://charmhub.io/grafana-k8s).

In order to add these dashboards to Grafana all that is required is to relate the two charms in the following way:

```bash
juju relate \
    grafana-agent-k8s:grafana-dashboard \
    grafana-k8s:grafana-dashboard
```

### Self metrics endpoint

```yaml
self-metrics-endpoint:
    interface: prometheus_scrape
```

> This Grafana Agent charm may forward information about its metrics endpoint and associated alert rules to a Prometheus charm over the `self-metrics-endpoint` relation using the [`prometheus_scrape`](https://charmhub.io/prometheus-k8s/libraries/prometheus_scrape) interface. In order for these metrics to be aggregated by the remote Prometheus charm all that is required is to relate the two charms as in:

Reviewer: ```suggestion
Grafana agent has its own metrics endpoint and the charm comes with associated alert rules, which can be used for self-monitoring. For this purpose the charm has the `self-metrics-endpoint` relation using the [`prometheus_scrape`](https://charmhub.io/prometheus-k8s/libraries/prometheus_scrape) interface. In order for these metrics to be aggregated by the remote Prometheus charm all that is required is to relate the two charms as in:
```

```bash
juju relate \
    grafana-agent-k8s:self-metrics-endpoint \
    prometheus-k8s:metrics-endpoint
```

## Requires

### Logging consumer

```yaml
  logging-consumer:
    interface: loki_push_api
```

[Loki](https://charmhub.io/loki-k8s) receives logs forwarded from Grafana Agent, aggregates and deduplicates them.
Grafana Agent Charmed Operator relates to Loki over the `loki_push_api` interface using the `logging-consumer` relation.

```bash
juju relate grafana-agent-k8s:logging-consumer loki-k8s
```

### Metrics endpoint

```yaml
  metrics-endpoint:
    interface: prometheus_scrape

> ```

Reviewer: I wonder if we should change the format a bit so things are more obvious and more copy-pasteable.
For example:

> ### Metrics endpoint
> Charms may forward information about their metrics endpoints and associated alert rules to the Grafana Agent charm over the `metrics-endpoint` relation using the [`prometheus_scrape`](https://charmhub.io/prometheus-k8s/libraries/prometheus_scrape) interface.
> 
> #### This charm
> ```yaml
> requires:
>   metrics-endpoint:
>     interface: prometheus_scrape
> ```
> 
> #### Remote charm
> 
> ```yaml
> provides:
>   metrics-endpoint:
>     interface: prometheus_scrape
> ```

etc.

Author: I like this format!

grafana-agent-k8s:grafana-dashboard      grafana:grafana-dashboard            grafana_dashboard        regular
grafana-agent-k8s:logging-provider       zinc-k8s:logging                     loki_push_api            regular
grafana-agent-k8s:self-metrics-endpoint  prometheus-k8s:metrics-endpoint      prometheus_scrape        regular
grafana:grafana                          grafana:grafana                      grafana_peers            peer
loki-k8s:logging                         grafana-agent-k8s:logging-consumer   loki_push_api            regular
prometheus-k8s:prometheus-peers          prometheus-k8s:prometheus-peers      prometheus_peers         peer
prometheus-k8s:receive-remote-write      grafana-agent-k8s:send-remote-write  prometheus_remote_write  regular
zinc-k8s:metrics-endpoint                grafana-agent-k8s:metrics-endpoint   prometheus_scrape        regular
```

## Deployment scenarios

```mermaid
flowchart LR
    subgraph COS Lite
        prometheus
        loki
        grafana
    end
    subgraph K8s cluster
        prometheus               ---      |remote_write| grafana-agent

>         prometheus               ---      |prometheus_scrape| grafana-agent

Reviewer: Iiuc, this line shouldn't be here:
- Both prom and agent have `metrics-endpoint` under requires
- The agent is "converting" scrape targets into remote write

Author: In fact this `prmetheus_scrape` is the interface used by `self-metrics-endpoint` relation, with this Prometheus can scrape Grafana Agent self metrics.

Should we show relation names too?

Author: Maybe we should not show self-metrics relations...

Reviewer: Yeah I think in the context of this deployment scenarios it is unnecessary.
I would certainly add the self monitoring relations in the bundle doc though.

Other kinds of self monitoring though could be worth mentioning here, but I think it would be a separate diagram. For example the resilience @rbarry82 was mentioning a while ago.

        - [Logging consumer](#logging-consumer)
            - [This charm](#this-charm-3)
            - [Remote charm](#remote-charm-3)
        - [Metrics endpoint](#metrics-endpoint)
            - [This charm](#this-charm-4)
            - [Remote charm](#remote-charm-4)
        - [Send remote write](#send-remote-write)
            - [This charm](#this-charm-5)
            - [Remote charm](#remote-charm-5)
    - [Deployment scenarios](#deployment-scenarios)

<!-- markdown-toc end -->

Grafana Agent Charmed Operator can be integrated with other charms through the following relations.

## Provides

### Logging provider

#### This charm

```yaml
provides:
  logging-provider:
    interface: loki_push_api
```

#### Remote charm

```yaml
requires:
  logging:
    interface: loki_push_api
```

Using this relation, Grafana Agent Charmed Operator can receive logs from any charm that supports the [`loki_push_api`](https://charmhub.io/loki-k8s/libraries/loki_push_api) relation interface (Those logs will later be forwarded to Loki using the `logging-consumer` relation).

The information exchanged through the `loki_push_api` interface can be broken down into two parts:

>   - Grafana Agent charm provides an [endpoint URL](https://grafana.com/docs/loki/latest/api/#post-lokiapiv1push) to receive log from charms running [Loki clients](https://grafana.com/docs/loki/latest/clients/).

Reviewer 2: ```suggestion
  - Grafana Agent charm provides an [endpoint URL](https://grafana.com/docs/loki/latest/api/#post-lokiapiv1push) to receive logs from charms running [Loki clients](https://grafana.com/docs/loki/latest/clients/).
```

>   - Grafana Agent charm may also receive alert rules which tell when to raise alerts. These rules are read from a directory named `loki_alert_rules`, if present at the top level, within the client charm's source (`src`) directory.

Reviewer 2: ```suggestion
  - Grafana Agent charm may also receive alert rules which tell it when to raise alerts. These rules are read from a directory named `loki_alert_rules`, if present at the top level, within the client charm's source (`src`) directory.
```

> For instance let's say that we need [Zinc Charmed Operator](https://charmhub.io/zinc-k8s) send logs to Grafana Agent.

Reviewer 2: ```suggestion
For instance let's say that we need [Zinc Charmed Operator](https://charmhub.io/zinc-k8s) to send logs to Grafana Agent.
```

> To do that we can use one of the "consumer" object offered by the `loki_push_api` lib and follow these steps.

Reviewer 2: ```suggestion
To do that we can use one of the "consumer" objects offered by the `loki_push_api` lib and follow these steps.
```

1. Specify the relation in the charm's metadata.yaml:

```yaml
requires:
  logging:
    interface: loki_push_api
```

2. Obtain the library from charmhub:

```shell
charmcraft fetch-lib charms.loki_k8s.v0.loki_push_api
```

3. Import the library and use it in your `src/charm.py`:

```python
from charms.loki_k8s.v0.loki_push_api import LogProxyConsumer
...
self._logging = LogProxyConsumer(self, relation_name="logging", log_files=logs_files)
```

4. After deploying Zinc charm relate it with `Grafana Agent`:

```bash
juju relate grafana-agent-k8s:logging-provider zinc-k8s:logging
```

Once the relation is established, Zinc charm can start sending logs to Grafana Agent charm.

### Grafana dashboard

#### This charm

```yaml
provides
  grafana-dashboard:
    interface: grafana_dashboard
```

#### Remote charm

```yaml
requires:
  grafana-dashboard:
    interface: grafana_dashboard
```

> Over the `grafana-dashboard` relation using the [`grafana-dashboard`](https://charmhub.io/grafana-k8s/libraries/grafana_dashboard) interface, this Grafana Agent charm also provides meaningful dashboards (placed in `src/grafana_dashboards`) about its metrics to be shown in a [Grafana Charm ](https://charmhub.io/grafana-k8s). These dashboards are placed

Reviewer 2: ```suggestion
Over the `grafana-dashboard` relation using the [`grafana-dashboard`](https://charmhub.io/grafana-k8s/libraries/grafana_dashboard) interface, this Grafana Agent charm also provides meaningful dashboards (placed in `src/grafana_dashboards`) about its metrics to be shown in a [Grafana Charm ](https://charmhub.io/grafana-k8s).
```

```bash
juju relate \
    grafana-agent-k8s:self-metrics-endpoint \
    prometheus-k8s:metrics-endpoint
```

## Requires

### Logging consumer

#### This charm
```yaml
requires:
  logging-consumer:
    interface: loki_push_api
```

#### Remote charm

```yaml
provides:
  logging:
    interface: loki_push_api
```

[Loki](https://charmhub.io/loki-k8s) receives logs forwarded by this Grafana Agent charm, aggregates and deduplicates them.
Grafana Agent Charmed Operator relates to Loki over the `loki_push_api` interface using the `logging-consumer` relation.

```bash
juju relate grafana-agent-k8s:logging-consumer loki-k8s
```

### Metrics endpoint

#### This charm

```yaml
requires:
  metrics-endpoint:
    interface: prometheus_scrape
```

#### Remote charm

```yaml
provides:
  metrics-endpoint:
    interface: prometheus_scrape
```

Charms may forward information about their metrics endpoints and associated alert rules to the Grafana Agent charm over the `metrics-endpoint` relation using the [`prometheus_scrape`](https://charmhub.io/prometheus-k8s/libraries/prometheus_scrape) interface.

```shell
juju relate grafana-agent-k8s:metrics-endpoint zinc-k8s
```

> Charms that seek to provide metrics endpoints and alert rules must do so using the provided [`prometheus_scrape`](https://charmhub.io/prometheus-k8s/libraries/prometheus_scrape) charm library.  This library by implementing the `metrics-endpoint` relation, not only ensures that scrape jobs and alert rules are forward to Grafana Agent but also that these are updated any time the metrics provider charm is upgraded. For example new alert rules may be added or old ones removed by updating and releasing a new version of the metrics provider charm. While it is safe to update alert rules as desired, care must be taken when updating scrape job specifications as this has the potential to break the continuity of the scraped metrics time series. In particular changing the following keys in the scrape job can break time series continuity

Reviewer 2: ```suggestion
Charms that seek to provide metrics endpoints and alert rules must do so by using the provided [`prometheus_scrape`](https://charmhub.io/prometheus-k8s/libraries/prometheus_scrape) charm library.  This library, by implementing the `metrics-endpoint` relation, not only ensures that scrape jobs and alert rules are forwarded to Grafana Agent but also that these are updated any time the metrics provider charm is upgraded. For example new alert rules may be added or old ones removed by updating and releasing a new version of the metrics provider charm. While it is safe to update alert rules as desired, care must be taken when updating scrape job specifications as this has the potential to break the continuity of the scraped metrics time series. In particular changing the following keys in the scrape job can break time series continuity
```

- `job_name`
- `relabel_configs`
- `metrics_relabel_configs`
- Any label set by `static_configs`

Evaluation of alert rules forwarded through the [`prometheus_scrape`](https://charmhub.io/prometheus-k8s/libraries/prometheus_scrape) interface are automatically limited to the charmed application that provided these rules. This ensures that alert rule evaluation is scoped down to the charm providing the rules.

### Send remote write

#### This charm

```yaml
requires:
  send-remote-write:
    interface: prometheus_remote_write
```

#### Remote charm

```yaml
provides:
  receive-remote-write:
    interface: prometheus_remote_write
```

Grafana Agent may forward client charms metrics and associated alert rules (that are received using the `metrics-endpoint` relation) to Prometheus using the [`prometheus_remote_write`](https://charmhub.io/prometheus-k8s/libraries/prometheus_remote_write) interface.
To do that you have to relate Grafana Agent with Prometheus:

```bash
juju relate grafana-agent-k8s:send-remote-write prometheus-k8s
```

> Once these relations are established, you can check that by running:

Reviewer 2: ```suggestion
Once these relations are established, you can check it by running:
```

The information exchanged through the `loki_push_api` interface can be broken down into two parts:

  - Grafana Agent charm provides an [endpoint URL](https://grafana.com/docs/loki/latest/api/#post-lokiapiv1push) to receive log from charms running [Loki clients](https://grafana.com/docs/loki/latest/clients/).
  - Grafana Agent charm may also receive alert rules which tell when to raise alerts. These rules are read from a directory named `loki_alert_rules`, if present at the top level, within the client charm's source (`src`) directory.

For instance let's say that we need [Zinc Charmed Operator](https://charmhub.io/zinc-k8s) send logs to Grafana Agent.
To do that we can use one of the "consumer" object offered by the `loki_push_api` lib and follow these steps.

1. Specify the relation in the charm's metadata.yaml:

```yaml
requires:
  logging:
    interface: loki_push_api
```

2. Obtain the library from charmhub:

```shell
charmcraft fetch-lib charms.loki_k8s.v0.loki_push_api
```

> 3. Import the library and use it in `src/cham.py`:

Reviewer: ```suggestion
3. Import the library and use it in your `src/charm.py`:
```
